### PR TITLE
feature(KtTable): initial implementation

### PR DESCRIPTION
**Pull request by Isokaeder**

### CODE REVIEW

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+	data: ROW[]
+	expandMode?: 'multi' | 'single'
+	getRowBehavior: GetRowBehavior<ROW>
+	hasDragAndDrop?: boolean
+	id: string
+	isSelectable?: boolean //{
+	// mode: 'single-page' | 'global'   // Consider negative selection for global case
+	// onSelectionUpdate: (updated: Record<string, boolean>) => void
+	// selectedRows: Ref<Record<string, boolean>>
+	//}
```
Comment (Isokaeder) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:58]: remove

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+	getRowBehavior: GetRowBehavior<ROW>
+} & {
+	expandMode: 'multi' | 'single' | null
+	hasDragAndDrop: boolean
+	id: string
+	isSelectable: boolean //{
+	// mode: 'single-page' | 'global'   // Consider negative selection for global case
+	// onSelectionUpdate: (updated: Record<string, boolean>) => void
+	// selectedRows: Ref<Record<string, boolean>>
+	//}
```
Comment (Isokaeder) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:167]: remove

File: packages/kotti-ui/source/kotti-table/table/local.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+		// always punish null
+		if (valueA === null) return map.A_IS_NULL
+		if (valueB === null) return map.B_IS_NULL
+
+		return valueA < valueB ? map.A_IS_SMALLER : map.B_IS_SMALLER
+	}
+}
+
+// export function getDateSorter<T extends Record<string, unknown>>(
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/local.ts:56]: Commented-out, also might be easy to replace without any date library, afaik we could just compare the dates with `>` or `<`

File: packages/kotti-ui/source/kotti-table/table/local.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	COLUMN_ID extends string,
+>(
+	data: Ref<ROW[]>,
+	ordering: Ref<KottiTable.Ordering<COLUMN_ID>[]>,
+	sortingFunctions: Record<
+		COLUMN_ID,
+		(mode: 'ascending' | 'descending') => (a: ROW, b: ROW) => SortInstruction
+	>,
+): void => {
+	const initialValue = clone(data.value)
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/local.ts:114]: Likely a bug when data is loaded remotely. Probably means this parameter must be split into two parts somehow

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+			toIndex > fromIndex ? toIndex - 1 : toIndex,
+			0,
+			fromColumnId,
+		)
+		return newOrder
+	}
+
+	const table = useVueTable<ROW>(
+		computed(() => ({
+			autoResetExpanded: true,
```
Comment (Isokaeder) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:398]: was added as an experiment that did not yield results. Remove

File: packages/kotti-ui/source/kotti-table/table/tanstack-table/index.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
+/**
+ * Code is taken from https://github.com/TanStack/table/discussions/4930
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/tanstack-table/index.ts:2]: might need to ask the author for approval to include this

File: packages/kotti-ui/source/kotti-table/table/types.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		REMOTE = 'REMOTE',
+	}
+
+	export type Translations = {
+		no: string
+		noItems: string
+		yes: string
+	}
+
+	export const paginationSchema = z.discriminatedUnion('type', [
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/types.ts:124]: Possibly to be removed?

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+import ToggleInner from '../../shared-components/toggle-inner/ToggleInner.vue'
+
+import { type TableContext, useProvideTableContext } from './context'
+import { useVueTable } from './tanstack-table'
+import { type GetRowBehavior, KottiTable } from './types'
+import { type ReactStyleUpdater, useComputedRef } from './use-computed-ref'
+
+export const EXPANSION_COLUMN_ID = 'internal-expand-column'
+export const SELECTION_COLUMN_ID = 'internal-select-column'
+export const ARRAY_START = 2
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:32]: Check if there are any odd usages of this left

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,652 @@
...
+	COLUMN_ID extends string,
+> = {
+	columns: KottiTable.Column<ROW, COLUMN_ID>[]
+	data: ROW[]
+	expandMode?: 'multi' | 'single'
+	getRowBehavior: GetRowBehavior<ROW>
+	hasDragAndDrop?: boolean
+	id: string
+	isSelectable?: boolean //{
+	// mode: 'single-page' | 'global'   // Consider negative selection for global case
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:55]: to be deleted(?)

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 10]
```diff
@@ -0,0 +1,652 @@
...
+> = {
+	columns: KottiTable.Column<ROW, COLUMN_ID>[]
+	data: ROW[]
+	getRowBehavior: GetRowBehavior<ROW>
+} & {
+	expandMode: 'multi' | 'single' | null
+	hasDragAndDrop: boolean
+	id: string
+	isSelectable: boolean //{
+	// mode: 'single-page' | 'global'   // Consider negative selection for global case
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:164]: to be removed (?)

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 11]
```diff
@@ -0,0 +1,652 @@
...
+				if (!newValue.includes(column.id)) {
+					newValue.push(column.id)
+				}
+			}
+			return newValue
+		},
+		value: ref([
+			EXPANSION_COLUMN_ID,
+			SELECTION_COLUMN_ID,
+			...params.value.columns.map(({ id }) => id),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:233]: Likely breaks reactivity

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 12]
```diff
@@ -0,0 +1,652 @@
...
+		]),
+	})
+
+	const pagination = useComputedRef<PaginationState>({
+		get: (value) => value,
+		set: (value) => {
+			const currentRow = value.pageIndex * value.pageSize
+			// @ts-expect-error TODO: fix pagination api
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:241]: `TODO`

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 13]
```diff
@@ -0,0 +1,652 @@
...
+			if (currentRow > params.value.pagination?.rowCount) {
+				return {
+					...value,
+					pageIndex: 0,
+				}
+			}
+			return value
+		},
+		value: ref(
+			params.value.pagination?.state ?? {
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:251]: Likely breaks reactivity

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+				pageIndex: 0,
+				pageSize: 10,
+			},
+		),
+	})
+
+	const selectedRows = useComputedRef<RowSelectionState>({
+		get: (value) => value,
+		set: (value) => value,
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:260]: May need to clear unknown rows, depending on use-case

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+
+			for (const id of columnIdSet.value) {
+				newVisibilityState[id] = !newSet.has(id)
+			}
+
+			return newVisibilityState
+		},
+		value: ref(
+			Object.fromEntries(
+				params.value.columns.map((column) => [column.id, true]),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:341]: Likely breaks reactivity

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 16]
```diff
@@ -0,0 +1,652 @@
...
+
+									return h(
+										'div',
+										{
+											class: {
+												'kt-table-expand': true,
+												yoco: true,
+											},
+											domProps: {
+												ariaDisabled: String(isDisabled),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:418]: Might need to be this

```suggestion
												'aria-disabled': String(isDisabled),
```

File: packages/kotti-ui/source/kotti-table/table/hooks.ts [hunk 17]
```diff
@@ -0,0 +1,652 @@
...
+												ariaExpanded: String(row.getIsExpanded()),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/hooks.ts:419]: Might need to be this

```suggestion
												'aria-expanded': String(row.getIsExpanded()),
```

File: packages/kotti-ui/source/kotti-table/table/column-helper.ts [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+			} as ResolvedDisplayType<OPTION>
+		}
+		case 'date':
+		case 'date-time': {
+			return {
+				align: 'left',
+				disableCellClick: false,
+				isNumeric: true,
+				render: (value: Date | null) =>
+					value === null ? null : dayjs(value).format(param.formatString),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/column-helper.ts:99]: Consider splitting this up/adding defaults somehow

File: packages/kotti-ui/source/kotti-table/table/column-helper.ts [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+			} as ResolvedDisplayType<OPTION>
+		}
+		case 'integer': {
+			return {
+				align: 'right',
+				disableCellClick: false,
+				isNumeric: true,
+				render: (value: number | null) =>
+					value === null ? null : String(Math.round(value)),
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/table/column-helper.ts:108]: Consider magnitude separators

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 20]
```diff
@@ -0,0 +1,651 @@
...
+					</th>
+					<th v-if="$scopedSlots['actions']" class="kt-table__actions-column" />
+				</tr>
+			</thead>
+			<tbody>
+				<tr v-if="isLoading" class="kt-table-row kt-table-row--is-loading">
+					<td :colSpan="tableColSpan">
+						<slot name="loading">
+							<div
+								v-for="key in 3"
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:43]: thanks, I hate it

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 21]
```diff
@@ -0,0 +1,651 @@
...
+							v-for="cell in row.cells"
+							:key="cell.key"
+							:class="cell.classes"
+							:data-test="cell.dataTest"
+							@dragenter.prevent
+							@dragleave.prevent
+							@dragover.prevent="(e) => handleCellDragOver(e, cell.columnId)"
+						>
+							<component
+								:is="cell.wrapComponent.component"
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:73]: This is re-usable for users right?

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 22]
```diff
@@ -0,0 +1,651 @@
...
+									:class="{
+										yoco: true,
+										'kt-table__action-icon': true,
+										'kt-table__action-icon--is-disabled':
+											row.isActionDisabled || action.isDisabled,
+									}"
+									:data-test="action.dataTest"
+									@click="
+										() => {
+											row.isActionDisabled || action.onClick()
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:104]:
```suggestion
											if (row.isActionDisabled) return
											
											action.onClick()
```

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 23]
```diff
@@ -0,0 +1,651 @@
...
+										}
+									"
+									v-text="action.icon"
+								/>
+								<!-- <slot name="actions" :row="row.original" :rowIndex="rowIndex" /> -->
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:109]: To be deleted(?)

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 24]
```diff
@@ -0,0 +1,651 @@
...
+						actions: behavior.actions ?? null,
+						cells: row.getVisibleCells().map((cell) => ({
+							classes: cell.column.columnDef.meta.cellClasses,
+							column: cell.column,
+							columnId: cell.column.id,
+							data:
+								typeof cell.column.columnDef.cell === 'string'
+									? cell.column.columnDef.cell
+									: cell.column.columnDef.cell?.({ ...cell.getContext() }) ??
+										'',
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:202]: Should this fall-back to empty string?

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 25]
```diff
@@ -0,0 +1,651 @@
...
+				const isLeftHalf = cursorX - elementX < elementWidth / 2
+				const targetIndex = columnIndex + (isLeftHalf ? 0 : 1)
+
+				tableContext.value.internal.setDropTargetColumnIndex(targetIndex)
+			},
+			handleDragLeave: (_event: DragEvent) => {
+				tableContext.value.internal.setDropTargetColumnIndex(null)
+			},
+			handleDrop: (event: DragEvent) => {
+				if (!isColumnMoveDataTransfer(event)) return
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:257]: Consider testing this

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 26]
```diff
@@ -0,0 +1,651 @@
...
+			table,
+			tableContext: computed(() => tableContext.value),
+			tableColSpan: computed(() => table.value.getAllFlatColumns().length),
+		}
+	},
+})
+</script>
+
+<style lang="scss">
+@keyframes fade-in-right {
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:330]: Rename

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 27]
```diff
@@ -0,0 +1,651 @@
...
+			margin-top: -0.8rem;
+			font-size: 0.8rem;
+			line-height: 0.8rem;
+			border: 1px solid var(--ui-02);
+			border-radius: var(--border-radius);
+
+			&--is-disabled {
+				color: var(--text-05);
+				cursor: not-allowed;
+				background-color: rgb(248 248 248 / 60%);
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:513]: Might want to consider putting comments here for the color this is modifying

File: packages/kotti-ui/source/kotti-table/KtTable.vue [hunk 28]
```diff
@@ -0,0 +1,651 @@
...
+		td {
+			text-align: center;
+		}
+	}
+
+	&--is-interactive {
+		cursor: pointer;
+
+		&:hover {
+			background-color: rgb(234 240 250 / 50%);
```
Comment (FlorianWendelborn) [inline packages/kotti-ui/source/kotti-table/KtTable.vue:576]: Same regarding comment

File: packages/documentation/pages/usage/components/table.vue [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+				},
+			]"
+		/>
+		<KtValueLabel
+			v-if="clickBehavior === 'simple-event'"
+			label="Row clicks"
+			:value="counter"
+		/>
+
+		<!-- TODO: should not reuse `id` -->
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:57]: `TODO`

File: packages/documentation/pages/usage/components/table.vue [hunk 30]
```diff
@@ -0,0 +1,520 @@
...
+	| 'isAlive'
+	| 'lifespan'
+	| 'name'
+	| 'preferredSound'
+	| 'purpose'
+	| 'randomDate'
+	| 'speed'
+	| 'worstEnemy'
+
+// | **Item**           | **Primary Purpose**     | **Speed**             | **Best Skill**        | **Worst Enemy**      | **Preferred Sound**  | **Lifespan**         | **Cuteness Level**   |
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:122]: table

File: packages/documentation/pages/usage/components/table.vue [hunk 31]
```diff
@@ -0,0 +1,520 @@
...
+			lifespan: string
+			name: string
+			preferredSound: string
+			purpose: string
+			someDate: string
+			speed: string | null
+			worstEnemy: string
+		}
+
+		// const exampleRow: TableRow = {
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:202]: dead

File: packages/documentation/pages/usage/components/table.vue [hunk 32]
```diff
@@ -0,0 +1,520 @@
...
+					}),
+					createColumn({
+						display: booleanDisplay,
+						getData: (row) => row.isAlive,
+						id: 'isAlive',
+						label: 'Aliveness',
+						width: '300px',
+					}),
+					createColumn({
+						// display: {
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:300]: dead

File: packages/documentation/pages/usage/components/table.vue [hunk 33]
```diff
@@ -0,0 +1,520 @@
...
+									{
+										icon: 'send',
+										onClick: () => {
+											// eslint-disable-next-line no-alert
+											window.alert('send icon clicked')
+										},
+									},
+								]
+							: undefined,
+					classes: [], //rowIndex % 2 === 0 ? ['everything-red'] : [],
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:399]: dead

File: packages/documentation/pages/usage/components/table.vue [hunk 34]
```diff
@@ -0,0 +1,520 @@
...
+				getTextSorter<TableRow>((row) => row.preferredSound, mode),
+			purpose: (mode) => getTextSorter<TableRow>((row) => row.purpose, mode),
+			randomDate: (mode) =>
+				getTextSorter<TableRow>((row) => row.someDate, mode),
+			speed: (mode) => getTextSorter<TableRow>((row) => row.speed, mode),
+			worstEnemy: (mode) =>
+				getTextSorter<TableRow>((row) => row.worstEnemy, mode),
+		})
+
+		// globalThis.table = tableHook
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:433]: should be commented out (?)

File: packages/documentation/pages/usage/components/table.vue [hunk 35]
```diff
@@ -0,0 +1,520 @@
...
+</script>
+
+<style>
+.everything-red {
+	color: red !important;
+	background-color: rosybrown !important;
+}
+</style>
+
+<style scoped>
```
Comment (FlorianWendelborn) [inline packages/documentation/pages/usage/components/table.vue:514]: dead

**Review by FlorianWendelborn — APPROVED** [review]
LGTM

**Pull request merged**